src/http.c

```diff
@@ -658,20 +658,37 @@ int auth_log_in(struct tunnel *tunnel)
 		get_value_from_response(res, "reqid=", reqid, 32);
 		get_value_from_response(res, "polid=", polid, 32);
 
-		if (cfg->otp[0] == '\0') {
-			read_password(cfg->pinentry, "otp",
-			              "Two-factor authentication token: ",
-			              cfg->otp, FIELD_SIZE);
+		if (cfg->otp[0] == '\0' &&
+		    strncmp(token, "ftm_push", 8) == 0 &&
+		    cfg->no_ftm_push == 0) {
+			/*
+			 * The server supports FTM push if `tokeninfo` is `ftm_push`,
+			 * but only try this if the OTP is not provided by the config
+			 * file or command line.
+			 */
+			snprintf(tokenparams, sizeof(tokenparams), "ftmpush=1");
+		} else {
 			if (cfg->otp[0] == '\0') {
-				log_error("No token specified\n");
-				return 0;
+				// Prompt for OTP token
+				read_password(cfg->pinentry, "otp",
+				              "Two-factor authentication token: ",
+				              cfg->otp, FIELD_SIZE);
+
+				if (cfg->otp[0] == '\0') {
+					log_error("No token specified\n");
+					return 0;
+				}
 			}
+
+			url_encode(tokenresponse, cfg->otp);
+			snprintf(tokenparams, sizeof(tokenparams),
+			         "code=%s&code2=",
+			         tokenresponse);
 		}
 
```

Comment:
Just want to make sure, what should happen if the server sends "ftpm_push" but `(cfg->otp[0] != '\0')`?

Reply:
If the config or command line specifies the OTP, I would expect that to have precedence. This is how I coded it (and manually tested it).

My reasoning; if you decided to use OTP explicitly, it doesn't seem to make sense to still use the FTM push, as both mechanisms are separate. (The forti vpn client gives you a GUI with the OTP field + submit button and a separate button for FTM push, when I tried it against my company's VPN.)